mopro-core/src/middleware/gpu_explorations/trapdoortech_zprize_msm/mod.rs

```diff
@@ -0,0 +1,164 @@
+mod local_msm;
+
+use ark_bls12_377_3::G1Affine;
+use ark_ec_3::AffineCurve;
+use ark_std::{One, Zero};
+use local_msm::{
+    edwards_from_neg_one_a, edwards_proj_to_affine, edwards_to_neg_one_a, edwards_to_sw,
+    multi_scalar_mul, sw_to_edwards, EdwardsAffine, ExEdwardsAffine,
+};
+use std::time::{Duration, Instant};
+
+use crate::middleware::gpu_explorations::utils::{benchmark::BenchmarkResult, preprocess};
+
+pub fn benchmark_msm<I>(
+    instances: I,
+    iterations: u32,
+) -> Result<Vec<Duration>, preprocess::HarnessError>
+where
+    I: Iterator<Item = preprocess::Instance>,
+{
+    let mut result_vec = Vec::new();
+
+    let mut ed_instances = vec![];
+    for instance in instances {
+        let points = &instance.0;
+
+        let mut ed_points = Vec::<ExEdwardsAffine>::new();
+        for p in points {
+            let ed_p = sw_to_edwards(EdwardsAffine {
+                x: p.x.clone(),
+                y: p.y.clone(),
+            });
+            let ed_p = edwards_to_neg_one_a(ed_p);
+            ed_points.push(ed_p);
+        }
+        ed_instances.push((ed_points, instance.1));
+    }
+
+    for instance in ed_instances {
+        let points = &instance.0;
+        let scalars = &instance.1;
+
+        let mut total_duration = Duration::ZERO;
+        for _i in 0..iterations {
+            let start = Instant::now();
+            let result = multi_scalar_mul(&points[..], &scalars[..]);
+            let result = edwards_from_neg_one_a(edwards_proj_to_affine(result));
+            let result = edwards_to_sw(result);
+
+            let _result = if result.x == <G1Affine as AffineCurve>::BaseField::zero()
+                && result.y == <G1Affine as AffineCurve>::BaseField::one()
+            {
+                G1Affine::new(result.x, result.y, true)
```

Comment:
This will improve the construction time (up to 99%)
```suggestion
                G1Affine::new_unchecked(result.x, result.y, true)
```
> ref: https://github.com/vimwitch/circom-compat/commit/91cd9ba69576269a7316a7512829c97395997588 and https://github.com/zkmopro/mopro/issues/25#issuecomment-2067880755

not sure if you have bottleneck here